File: node/m10-sdk/src/builders/transfer.ts
import { google, m10 } from "../../protobufs";
import type { AccountId, TxId } from "../ids";

import type { ITxnFilter } from "./txn_filter";
import { TxnFilter } from "./txn_filter";


export class StepBuilder {
    private _from: AccountId;
    private _to: AccountId;
    private _amount: number;
    private _metadata: google.protobuf.Any[];


    public constructor(from: AccountId, to: AccountId, amount: number) {
        this._from = from;
        this._to = to;
        this._amount = amount;
        this._metadata = [];
    }

    public metadata(value: google.protobuf.Any): StepBuilder {
        this._metadata.push(value);
        return this;
    }

    public custom_metadata(typeUrl: string, payload: Uint8Array): StepBuilder {
        this._metadata.push(new google.protobuf.Any({ type_url: typeUrl, value: payload }));
        return this;
    }

    public toTransferStep(): m10.sdk.transaction.TransferStep {
        return new m10.sdk.transaction.TransferStep({
            fromAccountId: this._from.toUint8Array(),
            toAccountId: this._to.toUint8Array(),
            amount: this._amount,
            metadata: this._metadata,
        });
    }
}


export class TransferBuilder {
    private _steps: StepBuilder[];

    public constructor() {
        this._steps = [];
    }

    public step(step: StepBuilder): TransferBuilder {
        this._steps.push(step);
        return this;
    }

    public toCreateTransfer(): m10.sdk.transaction.CreateTransfer {
        return new m10.sdk.transaction.CreateTransfer({
            transferSteps: this._steps.map((step) => step.toTransferStep()),
        });
    }
}

export interface ITransferFilter extends ITxnFilter {
    min: TxId;
    max: TxId;
    limit: number;
    includeChildAccounts: boolean;
    accountId: Option<AccountId>;
    contextId: Option<Uint8Array>;
}

export class TransferFilter extends TxnFilter {

    private _includeChildAccounts: boolean;
    private _accountId: Option<AccountId>;
    private _contextId: Option<Uint8Array>;


    public constructor(properties: ITransferFilter) {
        super(properties);

        this._includeChildAccounts = properties.includeChildAccounts;
        this._accountId = properties.accountId;
        this._contextId = properties.contextId;
    }


    public includeChildAccounts(enable: boolean): TransferFilter {
        this._includeChildAccounts = enable;
        return this;
    }

    public toListTransferRequest(): m10.sdk.transaction.ListTransferRequest {
        return new m10.sdk.transaction.ListTransferRequest({
            includeChildAccounts: this._includeChildAccounts,
            minTxId: this.getMinTxId(),
            maxTxId: this.getMaxTxId(),
            limit: this.getLimit(),
            accountId: this._accountId?.toUint8Array(),
            contextId: this._contextId,
        });
    }

    public limit(count: number): TransferFilter {
        super.limit(count);
        return this;
    }

    public minTxId(txId: LongNumber): TransferFilter {
        super.min(txId);
        return this;
    }

    public maxTxId(txId: LongNumber): TransferFilter {
        super.max(txId);
        return this;
    }

    public static byAccount(accountId: AccountId): TransferFilter {
        return new TransferFilter({
            min: 0,
            max: Number.MAX_SAFE_INTEGER,
            limit: TxnFilter.DEFAULT_TXN_LIMIT,
            includeChildAccounts: false,
            accountId: accountId,
            contextId: null,
        });
    }

    public static byContextId(contextId: Uint8Array): TransferFilter {
        return new TransferFilter({
            min: 0,
            max: Number.MAX_SAFE_INTEGER,
            limit: TxnFilter.DEFAULT_TXN_LIMIT,
            includeChildAccounts: false,
            accountId: null,
            contextId: contextId,
        });
    }
}

File: rust/sdk/src/builders/transfer.rs
use crate::account::AccountId;
use crate::builders::{TxnFilter, DEFAULT_TXN_LIMIT};
use crate::WithContext;
use core::convert::From;
use core::default::Default;
use m10_protos::prost::Any;
use m10_protos::sdk::list_transfer_request::Filter;
use m10_protos::{sdk, Metadata};

pub struct StepBuilder {
    from: AccountId,
    to: AccountId,
    amount: u64,
    metadata: Vec<Any>,
}

impl StepBuilder {
    pub fn new(from: AccountId, to: AccountId, amount: u64) -> Self {
        Self {
            from,
            to,
            amount,
            metadata: vec![],
        }
    }

    pub fn metadata(mut self, value: impl Metadata) -> Self {
        self.metadata.push(value.any());
        self
    }

    pub fn any_metadata(mut self, value: Any) -> Self {
        self.metadata.push(value);
        self
    }

    pub fn custom_metadata(
        mut self,
        type_url: impl Into<String>,
        payload: impl Into<Vec<u8>>,
    ) -> Self {
        self.metadata.push(Any {
            type_url: type_url.into(),
            value: payload.into(),
        });
        self
    }
}

impl From<StepBuilder> for sdk::TransferStep {
    fn from(step: StepBuilder) -> Self {
        Self {
            from_account_id: step.from.to_vec(),
            to_account_id: step.to.to_vec(),
            amount: step.amount,
            metadata: step.metadata,
        }
    }
}

#[derive(Default)]
pub struct TransferBuilder(Vec<StepBuilder>);

impl TransferBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn step(mut self, step: StepBuilder) -> Self {
        self.0.push(step);
        self
    }
}

impl From<TransferBuilder> for sdk::CreateTransfer {
    fn from(transfer: TransferBuilder) -> Self {
        Self {
            transfer_steps: transfer
                .0
                .into_iter()
                .map(sdk::TransferStep::from)
                .collect(),
        }
    }
}

pub struct TransferFilter {
    filter: Filter,
    include_child_accounts: bool,
}

impl TxnFilter<TransferFilter> {
    pub fn by_account(id: AccountId) -> Self {
        Self {
            filter: TransferFilter {
                filter: Filter::AccountId(id.to_vec()),
                include_child_accounts: false,
            },
            min: 0,
            max: u64::MAX,
            limit: DEFAULT_TXN_LIMIT,
        }
    }

    pub fn by_context_id(context_id: Vec<u8>) -> Self {
        Self {
            filter: TransferFilter {
                filter: Filter::ContextId(context_id),
                include_child_accounts: false,
            },
            min: 0,
            max: u64::MAX,
            limit: DEFAULT_TXN_LIMIT,
        }
    }

    pub fn include_child_accounts(mut self, enable: bool) -> Self {
        self.filter.include_child_accounts = enable;
        self
    }
}

impl From<TxnFilter<TransferFilter>> for sdk::ListTransferRequest {
    fn from(filter: TxnFilter<TransferFilter>) -> Self {
        sdk::ListTransferRequest {
            filter: Some(filter.filter.filter),
            min_tx_id: filter.min,
            max_tx_id: filter.max,
            include_child_accounts: filter.filter.include_child_accounts,
            limit: filter.limit,
        }
    }
}

impl WithContext for TransferBuilder {}
